okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/UrlSelectorImpl.java

```diff
@@ -155,12 +157,26 @@ private static String switchWsToHttp(String url) {
         return redirectTo(current, baseUrls.get((currentUrl.get() + 1) % baseUrls.size()));
     }
 
+    @Override
+    public Optional<HttpUrl> redirectToHash(HttpUrl current, Optional<String> nodePinValue) {
+        if (nodePinValue.isPresent()) {
+            Optional<HttpUrl> nextUrl = consistentHashRing.getNode(nodePinValue.get());
+            if (nextUrl.isPresent()) {
+                return redirectTo(current, nextUrl.get());
+            }
+        }
+
+        return redirectTo(current, baseUrls.get(currentUrl.get()));
+    }
+
     @Override
     public void markAsFailed(HttpUrl failedUrl) {
         if (useFailedUrlCache) {
             Optional<Integer> indexForFailedUrl = indexFor(failedUrl);
-            indexForFailedUrl.ifPresent(index ->
-                    failedUrls.put(baseUrls.get(index), UrlAvailability.FAILED)
+            indexForFailedUrl.ifPresent(index -> {
+                        failedUrls.put(baseUrls.get(index), UrlAvailability.FAILED);
+                        consistentHashRing.removeNode(baseUrls.get(index));
```

Comment:
I'm not sure this is right, the URLs never reenter the ring? I think you just want to use the failedUrls thing in redirectToHash to somehow select next node to try?